applications/FluidDynamicsApplication/python_scripts/navier_stokes_two_fluids_hydraulic_solver.py

```diff
@@ -0,0 +1,692 @@
+        if self.element_has_nodal_properties:
+            self._SetNodalProperties()
+
+        return materials_imported
+
+    def _SetNodalProperties(self):
+        # Get fluid 1 and 2 properties
+        properties_1 = self.main_model_part.Properties[1]
+        properties_2 = self.main_model_part.Properties[2]
+
+        rho_1 = properties_1.GetValue(KratosMultiphysics.DENSITY)
+        rho_2 = properties_2.GetValue(KratosMultiphysics.DENSITY)
+        mu_1 = properties_1.GetValue(KratosMultiphysics.DYNAMIC_VISCOSITY)
+        mu_2 = properties_2.GetValue(KratosMultiphysics.DYNAMIC_VISCOSITY)
+
+        # Check fluid 1 and 2 properties
+        if rho_1 <= 0.0:
+            raise Exception("DENSITY set to {0} in Properties {1}, positive number expected.".format(rho_1, properties_1.Id))
+        if rho_2 <= 0.0:
+            raise Exception("DENSITY set to {0} in Properties {1}, positive number expected.".format(rho_2, properties_2.Id))
+        if mu_1 <= 0.0:
+            raise Exception("DYNAMIC_VISCOSITY set to {0} in Properties {1}, positive number expected.".format(mu_1, properties_1.Id))
+        if mu_2 <= 0.0:
+            raise Exception("DYNAMIC_VISCOSITY set to {0} in Properties {1}, positive number expected.".format(mu_2, properties_2.Id))
+
+        # Transfer density and (dynamic) viscostity to the nodes
+        for node in self.main_model_part.Nodes:
+            if node.GetSolutionStepValue(self._levelset_variable) <= 0.0:
+                node.SetSolutionStepValue(KratosMultiphysics.DENSITY, rho_1)
+                node.SetSolutionStepValue(KratosMultiphysics.DYNAMIC_VISCOSITY, mu_1)
+            else:
+                node.SetSolutionStepValue(KratosMultiphysics.DENSITY, rho_2)
+                node.SetSolutionStepValue(KratosMultiphysics.DYNAMIC_VISCOSITY, mu_2)
+
+    def _GetRedistancingLinearSolver(self):
+        # A linear solver configured specifically for distance re-initialization process
+        if not hasattr(self, '_redistancing_linear_solver'):
+            self._redistancing_linear_solver = self._CreateLinearSolver() # TODO: add customized configuration
+        return self._redistancing_linear_solver
+
+    def _GetLevelsetLinearSolver(self):
+        # A linear solver configured specifically for the level-set convection process
+        if not hasattr(self, '_levelset_linear_solver'):
+            self._levelset_linear_solver = self._CreateLinearSolver() # TODO: add customized configuration
+        return self._levelset_linear_solver
+
+    #TODO: Leave it like this to remove it when its working
+    def _GetEulerianFmAleLinearSolver(self):
+        # A linear solver configured specifically for the level-set convection process
+        if not hasattr(self, '_eulerian_fm_ale_linear_solver'):
+            self._eulerian_fm_ale_linear_solver = self._CreateLinearSolver() # TODO: add customized configuration
+        return self._eulerian_fm_ale_linear_solver
```

Comment:
I think we can now have a unique linear solver for the convections.

Reply:
You are right!